File: TauResearch/ViewModels/VideoUploadViewModel.swift
// VideoUploadViewModel.swift
// Mirrors upload/viewmodel/VideoUploadViewModel.kt — drives the standalone "Upload" panel.

import Foundation
import Observation

struct UploadState: Equatable {
    var isUploading: Bool = false
    var progress: Float = 0
    var success: Bool = false
    var error: String?
}

@Observable
@MainActor
final class VideoUploadViewModel {

    private let repo = VideoUploadRepository()
    var state = UploadState()

    private var uploadTask: Task<Void, Never>?

    func startUpload(videoPath: String, gpxPath: String) {
        uploadTask?.cancel()
        state = UploadState(isUploading: true)
        uploadTask = Task { [weak self] in
            guard let self else { return }
            do {
                let videoUrls = repo.presignedVideoUrls()
                guard let gpxUrl = repo.presignedGpxUrl(),
                      let csvUrl = repo.presignedCsvUrl(),
                      !videoUrls.isEmpty else {
                    throw AppError.unexpected(underlying: nil)
                }
                let csvPath = (videoPath as NSString).deletingPathExtension + ".csv"
                _ = try await repo.uploadVideoInChunks(
                    videoPath: videoPath,
                    presignedUrls: videoUrls,
                    progress: { p in
                        Task { @MainActor in self.state.progress = p }
                    },
                    isCancelled: { Task.isCancelled }
                )
                try await repo.uploadGpxFile(path: gpxPath, presignedUrl: gpxUrl)
                try await repo.uploadCsvFile(path: csvPath, presignedUrl: csvUrl)
                state = UploadState(isUploading: false, progress: 1, success: true)
            } catch {
                state = UploadState(isUploading: false, error: error.localizedDescription)
            }
        }
    }

    func cancelUpload() {
        uploadTask?.cancel()
        state = UploadState(error: "Upload cancelled")
    }
}

File: app/src/main/java/com/example/cameraaccess/upload/viewmodel/VideoUploadViewModel.kt
package com.example.cameraaccess.upload.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.cameraaccess.data.model.ApiCache
import com.example.cameraaccess.upload.model.UploadState
import com.example.cameraaccess.upload.repository.VideoUploadRepository
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

class VideoUploadViewModel(application: Application) : AndroidViewModel(application) {

    private val apiCache = ApiCache(application)
    private val repository: VideoUploadRepository = VideoUploadRepository(apiCache)

    private val _uploadState = MutableStateFlow(UploadState())
    val uploadState = _uploadState.asStateFlow()

    private var uploadJob: Job? = null

    fun startUpload(videoPath: String, gpxPath: String) {
        uploadJob?.cancel()

        uploadJob = viewModelScope.launch {
            try {
                _uploadState.value = UploadState(isUploading = true)

                val videoUrls = repository.getPresignedVideoUrls()
                val gpxUrl = repository.getPresignedGpxUrl()
                val csvUrl = repository.getPresignedCsvUrl()
                val csvPath = videoPath.substringBeforeLast('.') + ".csv"

                if (videoUrls.isEmpty() || gpxUrl.isNullOrEmpty() || csvUrl.isNullOrEmpty()) {
                    throw Exception("Missing presigned URLs")
                }

                val etags = repository.uploadVideoInChunks(
                    videoPath = videoPath,
                    presignedUrls = videoUrls,
                    onProgress = {
                        _uploadState.value =
                            _uploadState.value.copy(progress = it)
                    },
                    isCancelled = { !isActive }
                )

                repository.uploadGpxFile(
                    gpxPath = gpxPath,
                    presignedUrl = gpxUrl
                )

                repository.uploadCsvFile(
                    csvPath = csvPath,
                    presignedUrl = csvUrl
                )

                _uploadState.value = UploadState(
                    isUploading = false,
                    progress = 1f,
                    success = true
                )

            } catch (e: Exception) {
                _uploadState.value = UploadState(
                    isUploading = false,
                    error = e.message
                )
            }
        }
    }

    fun cancelUpload() {
        uploadJob?.cancel()
        _uploadState.value = UploadState(error = "Upload cancelled")
    }
}